### DISCUSSION TITLE
Libraries

### DISCUSSION THREAD

**Discussion by R3dst0ne** [logisim-evolution/logisim-evolution#690] [General]
Since the early responses missed what this should be about (probably due to my bad initial description). I decided to clarify it.
1.  - What kind of libraries/components do we want to be included with Logisim-evolution?
     - What are the requirements to be included?
     
2.  - Should we start a list with libraries that are not included?
        - What are the requirements to be listed?
        - Only 3rd party or  also ones created here?
     
3.  - (Check if the currently included libraries satisfy the in point 1 defined conditions.)

4. Small additional points: 
    - The Documentation for libraries needs to be updated.
    - As mentioned in #600, the placement of the [example library](https://github.com/logisim-evolution/logisim-evolution/tree/master/src/main/java/com/cburch/gray) is suboptimal and it needs to be updated.
    - Most external JAR-Libraries are going to break in the next releas. (because of #648)
    
What are your idears/opinions on these points?

**Comment by BFH-ktt1** [comment]
So there are many reasons for the different libraries:
1) TTL-Library is merged from logisim-ITA and extended
2) IO-extra library is also merged from logisim-ITA. Why I called it extra, as I did not want it at the moment to move it into IO, but we can merge it in IO.
3) SoC, although very specialized, has it's roots deep down in the simulation engine of logisim, and hence will not be easy to move out.
4) TCL is also used for the Modelsim/Questasim co-simulation.
As for the two PLA's, one is the one made by the logisim-ITA team, and the other is made by Kevin Walsh.

So I do not see a need to move them out. A better way to go about it is to adapt the template "default.templ" that isloaded at startup. Here you can specify whick "build-in" libraries are visible:
```
 <lib name="0" desc="#Wiring" />
 <lib name="1" desc="#Gates" />
 <lib name="2" desc="#Plexers" />
 <lib name="3" desc="#Arithmetic" />
 <lib name="4" desc="#Memory" />
 <lib name="5" desc="#I/O" />
 <lib name="A" desc="#TTL" />
 <lib name="7" desc="#TCL" />
 <lib name="8" desc="#Base" />
 <lib name="9" desc="#BFH-Praktika" />
 <lib name="B" desc="#Input/Output-Extra" />
 <lib name="C" desc="#Soc" />
```

by just removing some lines, the libraries are not any more visible, but can be "made" visible again in logisim.
The same also goes for example for the top-toolbar, and mappings. So everybody can make their "default template" for their purposes.
What do you think?

**Comment by R3dst0ne** [comment]


**Reply by MarcinOrlowski** [reply]
> The SoC is not a standard library, but is included.
> TTL is great, but should not belong to the standard.

May I ask how is "standard library" defined?

> BFH is neat to have, but not really needed and only features two components.

Are you saying that the more elements makes it harder to find the one you need? Or why is less elements available better?

>  Because it features real world components that can easily be replicated with gates from the other standard libraries.

Well, almost everything can (except for flip-flops which will not work properly with current simulation engine :)

Anyway, I do agree that the libraries may need organizing. Templates may be good way but it's completely user-unfriendly. User should be able to i.e. hide certain groups, components using UI. Or maybe star some as favorite if s/he uses them a lot. Not to mention component tree needs a filter feature. Do I miss anything?

**Reply by R3dst0ne** [reply]
- SoC lies in its own [folder/namespace](https://github.com/logisim-evolution/logisim-evolution/tree/master/src/main/java/com/cburch/logisim/soc) and not in [sdt](https://github.com/logisim-evolution/logisim-evolution/tree/master/src/main/java/com/cburch/logisim/std) where all the others are.

- > why is less elements available better?

     You misunderstood me. :sweat_smile: I was saying it is not really needed and having a build in library for only two components is pointless espacialy when they feature so little functionality.

- > User should be able to i.e. hide certain groups, components using UI. 

    Under `Project` you can _load_ or _unload_ Libraries. 
    
    > Or maybe star some as favorite if s/he uses them a lot.

    Under `Project > Options > Toolbar` you can edit your toolbar.
    
    There only needs to be a way to save these changes in a `custom.templ` file.
    The prevered theme (_Look and Feel_) and backround colours could also be saved there.

**Comment by R3dst0ne** [comment]
My proposal/opinion:
1. Libraries that feature basic components that are widely used.
    - They should be useful for the majority of Logisim-evolution users.
    - only basic components or things that would otherwise be really hard to implement.
    E.g.  ALU-Component :thumbsdown: | I/O-Components :thumbsup:
    - No duplicates. E.g. we don't need a library that includes multiple LEDs that only differ in color.
    - FPGA support (if doable)
    - Only JAR-Libraries.
2. I think everybody could profit from such a list.
    - the only requirements I can think of is that they have to serve some kind of purpose, work and are compatible to the newest releas.
    - I think it would be awesome if this community could maintain libraries that are no longer maintained by their creators.
    E.g. like @kevinawalsh did [here](https://github.com/kevinawalsh/logisim-evolution/tree/holycross/contrib).
3. TTL, TCL & SoC wouldn't match my conditions
    - TCL & SoC should stay because of the reasons @BFH-ktt1 mentioned
    - TTL could be moved to the Library list, because its components are duplicates and could easily be manually implemented.
4.
   - I think it would be good if older libraries would still work in the next releas.
     On the other hand, @MarcinOrlowski did a view great improvements, and it would be better if everybody adopted them.
   - Providing an example how external library should be organized/function could help.
      @MarcinOrlowski since you did the most changes to the libraries. Could you update the [Gray library](https://github.com/logisim-evolution/logisim-evolution/tree/develop/src/main/java/com/cburch/gray), so that it can be used as an example?

**Reply by MarcinOrlowski** [reply]
> Most external JAR-Libraries are going to break in the next releas. (because of #648)

I am not sure #648 would nuke external libs. There's nothing there that could affect them (the only thing I can can think of is implementation of `getName()` that currently looks for `_ID` in the library class. But this was formerly returning hardcoded string so JARs should be safe here. But what could bite them is #638 as it removes old constructors for `FactoryDescription` objects. The fix is to restore old API which I will do soon (unless there's a reason for not doing so anyway?). In fact there's no reason I can think of now to break backward compatibility as there's no improvement that would justify that or couldn't be done w/o it.

> I think it would be awesome if this community could maintain libraries that are no longer maintained by their creators.

Pushing any library external w/o offering easy way of discovering it by the users (which usually is done by i.e. central repository or built-in manager feature etc) usually equals killing such library. This is still sort-of niche project so reducing traffic/interest to 3rd party elements is actually not what is desired as their maintainers will quickly loose motivation to take care and drop it.

---

But I think this all exposes more serious problem here - one thing is I should have be more foreseeing, but also the #638 should not be merged with backward incompatible change or/and should be caught during code review (or tests :).

So my proposal here is to hold right now and plan ahead. I created #695 to talk ROADMAP.

**Reply by MarcinOrlowski** [reply]
As I started fixing code style I realized that this can also cause backward incompatibilities as some methods will be renamed so JARs may no longer work. So here is the question - how many libraries is out there? Is this a problem? From one hand, fix should be relatively simple - correct the spelling of some methods, and voila. From the other hand, it'd be also relatively simple to keep old API names around as bridge too. And I think this is the best solution as it gives more time to correct external libs, while not blocking code cleanup. @BFH-ktt1?

**Reply by BFH-ktt1** [reply]
@MarcinOrlowski : As far as I know is the jar-library support still broken. I noticed some pushes recently by @kevinawalsh concerning the jar libraries. As I did not jet merge all of his nice improvements it is difficult to make a real decision. I would go for a new form of jar libraries that works, but I really have no idea how many libraries are out there.
